### PR TITLE
Added Mso library required for ReactHost

### PR DESCRIPTION
**Pull request by vmoroz**
The new ReactHost implementation depends on code that we currently have in Office internally known as Mso library. Recently we have published an initial set of Mso liblets as github.com/microsoft/Mso project, but this one is missing DispatchQueues and Mso::Future required by ReactHost.

In this change we add a new simplified version of DispatchQueue and existing Mso::Future directly to react-native-windows project. This is a temporary code until we put all required code into the Mso repo and implement deployment mechanisms. But it should unblock us with the ReactHost implementation.

The new DispatchQueue implementation is missing timers and waiters, but it has a much more simplified code that also can do more:
* The new IDispatchQueueService can 
  * schedule and invoke tasks
  * shutdown queue
  * do task batching
  * support Queue Local Values (like TLS for threads)
* All platforms specific work is done by IDispatchQueueScheduler implementations

The change also includes implementation of [noexcept2] attribute for IDL files that will be replaced in future with the standard [noexcept] attribute when we start using new Windows SDK.
 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3905)

### CODE REVIEW

**Comment by NikoAri** [comment]
> // Copyright (c) Microsoft Corporation. All rights reserved.

(nit) should we follow standard and make this file "AbiViewManager.h" or "abiViewManager.h"

---
Refers to: vnext/Microsoft.ReactNative/ABIViewManager.h:1 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

File: vnext/Microsoft.ReactNative/IJSValueReader.idl [hunk 1]
```diff
@@ -17,12 +19,12 @@ namespace Microsoft.ReactNative {
...
   [webhosthidden]
   interface IJSValueReader {
-    JSValueType ValueType { get; };
-    Boolean GetNextObjectProperty(out String propertyName);
-    Boolean GetNextArrayItem();
-    String GetString();
-    Boolean GetBoolean();
-    Int64 GetInt64();
-    Double GetDouble();
+    [noexcept2] JSValueType ValueType { get; };
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/IJSValueReader.idl]: >noexcept2 [](start = 5, length = 9)

why noexcept2 instead of just noexcept?

**Comment by NickGerleman** [comment]
>   ArrayView &operator=(const ArrayView &other) noexcept {

I think all of these got formatted with RNW style when copied into this source tree. This is going to make diffing painful if we try to reconcile the two repos.

---
Refers to: vnext/Mso/future/details/arrayView.h:31 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
> struct LooperScheduler : Mso::UnknownObject<Mso::RefCountStrategy::WeakRef, IDispatchQueueScheduler> {

Most of the new work is in this directory right? I would really appreciate it if in the future we could make it easier to tell what has been changed. E.g. multiple PRs or at least multiple commits. This would be much less of an issue though if we did this in the Mso repo itself.

---
Refers to: vnext/Mso/src/dispatchQueue/looperScheduler.cpp:10 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
> /*static*/ Mso::CntPtr<IDispatchQueueScheduler> DispatchQueueStatic::MakeLooperScheduler() noexcept {

Where did we rename TCntPtr? Was this just for this repo or did this happen in Mso or devmain?

---
Refers to: vnext/Mso/src/dispatchQueue/looperScheduler.cpp:100 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
> struct LooperScheduler : Mso::UnknownObject<Mso::RefCountStrategy::WeakRef, IDispatchQueueScheduler> {

Similarly, a lot of this looks like same as what's already in DispatchQueue in devmain, but some parts look new. Could you provide a way to diff this?

---
In reply to: [575861025](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575861025) [](ancestors = 575861025)

---
Refers to: vnext/Mso/src/dispatchQueue/looperScheduler.cpp:10 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
> #include "winrt/Windows.ApplicationModel.Core.h"

nit: angle brackets for external headers

---
Refers to: vnext/Mso/src/dispatchQueue/mainUIScheduler_winrt.cpp:8 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
>     <ClCompile Include="$(MSBuildThisFileDirectory)MemoryMappedBuffer.cpp" />

IIRC this will break the build when using OSS_RN or unpatched RN

---
Refers to: vnext/Shared/Shared.vcxitems.filters:10 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = True)

File: vnext/Mso/dispatchQueue/readMe.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# dispatchQueue liblet
```
Comment (NickGerleman) [inline vnext/Mso/dispatchQueue/readMe.md]: >dispatchQueue [](start = 2, length = 13)

nit: this file is conventionally README with all caps

**Comment by NickGerleman** [comment]
> #include "winrt/Windows.UI.Composition.h"

nit: angle brackets

---
Refers to: vnext/ReactUWP/Pch/pch.h:27 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
> #include "winrt/Windows.ApplicationModel.Core.h"

same for other files

---
In reply to: [575861334](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575861334) [](ancestors = 575861334)

---
Refers to: vnext/Mso/src/dispatchQueue/mainUIScheduler_winrt.cpp:8 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
> 

I really appreciate that you removed the pch from public headers. On a thread we were on recently, someone mentioned the strategy of removing the #pragma once, then failing the build if the pch is double included. Are we able to do that cleanly here?

---
Refers to: vnext/ReactUWP/Pch/pch.h:5 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

File: vnext/Mso/dispatchQueue/readMe.md (hunk 2 above)
Comment (NickGerleman) [reply vnext/Mso/dispatchQueue/readMe.md]: same for future

---
In reply to: [368202671](https://github.com/microsoft/react-native-windows/pull/3905#discussion_r368202671) [](ancestors = 368202671)

File: vnext/Microsoft.ReactNative/IJSValueReader.idl (hunk 1 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/IJSValueReader.idl]: The noexcept is the reserved keyword in future versions of MIDL. We cannot use it yet. Though, the same functionality can be provided by a custom attribute for now. Thus, we should have a different name and I chose the noexcept2.

---
In reply to: [368188593](https://github.com/microsoft/react-native-windows/pull/3905#discussion_r368188593) [](ancestors = 368188593)

**Comment by vmoroz** [comment]
> // Copyright (c) Microsoft Corporation. All rights reserved.

The Mso library uses the standard Microsoft header as required by Microsoft legal guidelines. These are the same header that you put in the [https://github.com/microsoft/Mso](https://github.com/microsoft/Mso). The react-native-windows uses non-official header for some reason.

---
In reply to: [575845103](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575845103) [](ancestors = 575845103)

---
Refers to: vnext/Microsoft.ReactNative/ABIViewManager.h:1 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by vmoroz** [comment]
>   ArrayView &operator=(const ArrayView &other) noexcept {

This is a temporary measure because all source code in the react-native-windows must be formatted using the same rules.
In future we will remove this code from the repo in favor of deployed one. There will be no need to reconcile.

---
In reply to: [575860861](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575860861) [](ancestors = 575860861)

---
Refers to: vnext/Mso/future/details/arrayView.h:31 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by vmoroz** [comment]
> struct LooperScheduler : Mso::UnknownObject<Mso::RefCountStrategy::WeakRef, IDispatchQueueScheduler> {

The dispatch queue implementation here is all new. Our current implementation due to to so many dependencies is almost impossible to publish. I have 'borrowed' some ideas from my current dispatch queue implementation in devmain and previous version done by Tudor (e.g the use of schedulers), but otherwise there is nothing to compare. This is in some sense the dispatch queues 2.0 which is if proven to work and do what we need, we are going to push back to Office devmain at some point. But, it is still missing a lot of functionality needed by Office code and it may be not that soon.

---
In reply to: [575861289](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575861289) [](ancestors = 575861289,575861025)

---
Refers to: vnext/Mso/src/dispatchQueue/looperScheduler.cpp:10 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by vmoroz** [comment]
> /*static*/ Mso::CntPtr<IDispatchQueueScheduler> DispatchQueueStatic::MakeLooperScheduler() noexcept {

This is just here. Using T as a class prefix is unconventional and I am trying to improve it here. All improvements that I have done in this PR to Mso are going to be ported to Mos github repo first, and then migrate to devmain.

---
In reply to: [575861149](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575861149) [](ancestors = 575861149)

---
Refers to: vnext/Mso/src/dispatchQueue/looperScheduler.cpp:100 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by vmoroz** [comment]
> #include "winrt/Windows.ApplicationModel.Core.h"

Angle brackets must be used for system files. In this case these are includes that are part of the project that includes this shared library. Let's discuss it offline. I can get back to this code and change it later.

---
In reply to: [575861893](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575861893) [](ancestors = 575861893,575861334)

---
Refers to: vnext/Mso/src/dispatchQueue/mainUIScheduler_winrt.cpp:8 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by vmoroz** [comment]
This is a temporary measure to get the ReactInstance management sooner. We will iterate on it and the Mso repo to get rid of code copy in favor of proper deployment model. I just must focus on the ReactInstance changes first.

---
In reply to: [344905252](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-344905252) [](ancestors = 344905252)

**Comment by vmoroz** [comment]
> 

I am going to add it in another PR.

---
In reply to: [575862127](https://github.com/microsoft/react-native-windows/pull/3905#issuecomment-575862127) [](ancestors = 575862127)

---
Refers to: vnext/ReactUWP/Pch/pch.h:5 in c3ee2ff. [](commit_id = c3ee2ff9a6eb70d59e452d1eb81ce77e9bded43a, deletion_comment = False)

**Comment by NickGerleman** [comment]
Not for right now, but at some point it might make sense to try to combine the ReactUWP and Microsoft.ReactNative PCH into one artifact.

**Pull request merged**